src/Npgsql/VolatileResourceManager.cs

```diff
@@ -224,6 +232,38 @@ public void InDoubt(Enlistment enlistment)
             }
         }
 
+        public void WaitIfRequired()
+        {
+            try
+            {
+                if (_secondPhaseEndedOrNotNeedingWait || _isRMCall.Value ||
+                    _transaction?.TransactionInformation.Status == System.Transactions.TransactionStatus.Active)
+                    return;
```

Comment:
Is this an optimization to avoid waiting on the semaphore if we don't need to (even when it's not taken)? If so, I do understand the motivation for this optimization - `WaitIfRequired()` gets called from StartUserAction, which makes it a hot path - just trying to make sure I'm not missing something else.